tools/pubsys/src/aws/publish_ami/mod.rs

```diff
@@ -0,0 +1,459 @@
+            }
+        );
+        ensure!(
+            images.len() == 1,
+            error::MultipleImages {
+                region: region.name(),
+                images: images
+                    .into_iter()
+                    .map(|i| i.image_id.unwrap_or_else(|| "<missing>".to_string()))
+                    .collect::<Vec<_>>()
+            }
+        );
+        let image = images.remove(0);
+
+        // Look into the block device mappings for snapshots.
+        let bdms = image
+            .block_device_mappings
+            .context(error::MissingInResponse {
+                request_type: "DescribeImages",
+                missing: "block_device_mappings",
+            })?;
+        ensure!(
+            !bdms.is_empty(),
+            error::MissingInResponse {
+                request_type: "DescribeImages",
+                missing: "non-empty block_device_mappings"
+            }
+        );
+        let mut snapshot_ids = Vec::with_capacity(bdms.len());
+        for bdm in bdms {
+            let ebs = bdm.ebs.context(error::MissingInResponse {
+                request_type: "DescribeImages",
+                missing: "ebs in block_device_mappings",
+            })?;
+            let snapshot_id = ebs.snapshot_id.context(error::MissingInResponse {
+                request_type: "DescribeImages",
+                missing: "snapshot_id in block_device_mappings.ebs",
+            })?;
+            snapshot_ids.push(snapshot_id);
+        }
+        snapshots.insert(region, snapshot_ids);
+    }
+
+    Ok(snapshots)
+}
+
+/// Modify snapshot attributes to make them public/private as requested.
+async fn modify_snapshots(
+    snapshots: &HashMap<Region, Vec<String>>,
+    clients: &HashMap<Region, Ec2Client>,
+    operation: String,
+) -> Result<()> {
+    // Build requests to modify snapshot attributes.
+    let mut modify_snapshot_requests = Vec::new();
+    for (region, snapshot_ids) in snapshots {
+        for snapshot_id in snapshot_ids {
+            let ec2_client = &clients[region];
+            let modify_snapshot_request = ModifySnapshotAttributeRequest {
+                attribute: Some("createVolumePermission".to_string()),
+                group_names: Some(vec!["all".to_string()]),
+                operation_type: Some(operation.clone()),
+                snapshot_id: snapshot_id.clone(),
+                ..Default::default()
+            };
+            let modify_snapshot_future =
+                ec2_client.modify_snapshot_attribute(modify_snapshot_request);
+
+            // Store the region and snapshot ID so we can include it in errors
+            let info_future = ready((region.name().to_string(), snapshot_id.clone()));
+            modify_snapshot_requests.push(join(info_future, modify_snapshot_future));
+        }
+    }
+
+    // Send requests in parallel and wait for responses, collecting results into a list.
+    let request_stream = stream::iter(modify_snapshot_requests).buffer_unordered(4);
```

Comment:
There's a much longer comment explaining this in `ami/mod.rs`, because it matters a lot more there; here, we just need to turn the stream iter into something we can wait on, and you have to pick a number, and it really doesn't matter because these requests are so quick.  :|